File: graph/graph.py
from langgraph.graph import StateGraph, START, END

from graph.state import ResearchState
from graph.nodes import (
    elicit_domain,
    research_papers,
    extract_triples,
    persist_to_graph,
    reason_datalog,
    reason_asp_conflicts,
    synthesize_response,
)
from graph.router import should_run_asp


def build_graph():
    """Build and compile the neuro-symbolic research graph."""
    builder = StateGraph(ResearchState)

    # Register nodes
    builder.add_node("elicit", elicit_domain)
    builder.add_node("research", research_papers)
    builder.add_node("extract_triples", extract_triples)
    builder.add_node("persist", persist_to_graph)
    builder.add_node("reason_datalog", reason_datalog)
    builder.add_node("reason_asp_conflicts", reason_asp_conflicts)
    builder.add_node("synthesize", synthesize_response)

    # Static edges
    builder.add_edge(START, "elicit")
    builder.add_edge("elicit", "research")
    builder.add_edge("research", "extract_triples")
    builder.add_edge("extract_triples", "persist")
    builder.add_edge("persist", "reason_datalog")
    # (persist → reason_datalog already added above)

    # Conditional: after Datalog, go to ASP conflict resolution OR direct synthesis
    builder.add_conditional_edges(
        "reason_datalog",
        should_run_asp,
        ["reason_asp_conflicts", "synthesize"],
    )

    builder.add_edge("reason_asp_conflicts", "synthesize")
    builder.add_edge("synthesize", END)

    return builder.compile()

File: src/graph/graph.ts
import { StateGraph, START, END } from "@langchain/langgraph";
import { ResearchState } from "./state.js";
import {
  elicitNode,
  researchNode,
  extractTriplesNode,
  persistNode,
  reasonDatalogNode,
  reasonAspConflictsNode,
  synthesizeNode,
} from "./nodes.js";
import { shouldRunAsp } from "./router.js";

export function buildResearchGraph() {
  return new StateGraph(ResearchState)
    .addNode("elicit", elicitNode)
    .addNode("research", researchNode)
    .addNode("extract_triples", extractTriplesNode)
    .addNode("persist", persistNode)
    .addNode("reason_datalog", reasonDatalogNode)
    .addNode("reason_asp_conflicts", reasonAspConflictsNode)
    .addNode("synthesize", synthesizeNode)
    .addEdge(START, "elicit")
    .addEdge("elicit", "research")
    .addEdge("research", "extract_triples")
    .addEdge("extract_triples", "persist")
    .addEdge("persist", "reason_datalog")
    .addConditionalEdges("reason_datalog", shouldRunAsp, [
      "reason_asp_conflicts",
      "synthesize",
    ])
    .addEdge("reason_asp_conflicts", "synthesize")
    .addEdge("synthesize", END)
    .compile();
}

let _graph: ReturnType<typeof buildResearchGraph> | undefined;

function getGraph() {
  if (!_graph) _graph = buildResearchGraph();
  return _graph;
}

/** Run the full research pipeline, optionally streaming progress events */
export async function runResearchGraph(
  query: string,
  onEvent?: (event: string) => void,
): Promise<{ answer: string; events: string[] }> {
  const graph = getGraph();

  if (onEvent) {
    const stream = graph.streamEvents(
      { query },
      { version: "v2", streamMode: "custom" },
    );
    for await (const event of stream) {
      if (event.event === "on_custom_event") {
        onEvent(event.data as string);
      }
    }
    // After streaming, get final state
    const result = await graph.invoke({ query });
    return { answer: result.answer, events: result.events };
  }

  const result = await graph.invoke({ query });
  return { answer: result.answer, events: result.events };
}
